L58-python-async-api.md

```diff
@@ -0,0 +1,1673 @@
+handler itself, otherwise the user needs to code a dedicated iterator class. For
+larger application, it would be a pain to create such iterator class for each
+streaming handler.
+
+euroelessar@ points out several great reasons for the new streaming API:
+* Enable `with` statement for streaming API, which allows clean resource management;
+* Easier to buffer outbound messages when peer pushes back (e.g. `yield` will
+  stop the execution of current coroutine, but `call.write` can be scheduled in
+  event loop);
+* Enable application to have both pending read and pending write in one
+  coroutine, and react on the first finisher (with iterator syntax this will be
+  quite hard to get it right).
+
+So, this gRFC introduces a new pattern of streaming API that reads/writes
+message to peer with explicit call.
+
+#### Snippet For Reader / Writer Streaming API
+
+```Python
+### Client side
+with stub.StreamingHi() as streaming_call:
+  request = echo_pb2.EchoRequest(message="ping")
+  await streaming_call.write(request)
+  response = await streaming_call.read()
+  while response:  # or response is not grpc.aio.EOF
+      process(response)
+      response = await streaming_call.read()
+
+### Server side
+class AsyncGreeter(helloworld_pb2_grpc.GreeterServicer):
+
+    async def StreamingHi(self,
+                          unused_request_iterator,
+                          context: grpc.aio.ServicerContext
+            ) -> None:
+        bootstrap_request = await context.read()
+        ...(bootstrap_request)
+        while ...:
+            response = ...
+            await context.write(response)
+```
+
+#### Snippet For Async Iterator Streaming API
+
+```Python
+### Client side
+
+async def async_request_iterator() -> HelloRequest:
+    for request in [...]:
+        yield response
+
+# No await needed, the response_iterator is grpc.aio.Call
+response_iterator = stub.StreamingHi(async_request_iterator())
+async for response in response_iterator:
+    process(response)
+
+### Server side
+
+class Greeter(helloworld_pb2_grpc.GreeterServicer):
+    async def StreamingHi(self, request_iterator, context):
+        async for request in request_iterator:
+            yield response
+```
+
+#### Co-existence Of New And Current Streaming API
+
+Existing API still has advantage over simpler use cases of gRPC, and iterator
+syntax is Pythonic to use. Also, for backward compatibility concern,
+iterator-based API needs to stay. So, both new and current streaming API needs
+to be included in the surface API.
+
+To keep the function signature stable, the new streaming API only requires
+enhancement to the context object (`grpc.aio.Call` on client-side, and
+`grpc.aio.ServicerContext` on the server-side).
+
+Notably, both new and current streaming API allows read and write messages
+simultaneously in different thread or coroutine. However, **the application should**
+**not mixing new and current streaming API at the same time**. Reading or writing to
```

Comment:
Even though mixing the two APIs is not supported, it might make sense to implement in a way that this is possible where the implementation would be simple enough (e.g.   servicer.read() could just call anext(request_iterator) and that's it?).